File: programming_skills/54_spiral_matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let row = 0, col = 0, drow = 0, dcol = 1;
    const spiralMat = []

    for (let i = 0; i < matrix.length * matrix[0].length; i++) {
        spiralMat.push(matrix[row][col]);
        matrix[row][col] = null;
        if (row + drow >= matrix.length ||
            col + dcol >= matrix[0].length ||
            row + drow < 0 ||
            col + dcol < 0 ||
            matrix[row + drow][col + dcol] === null
        ) {
            [drow, dcol] = [dcol, -drow];
        }
        row += drow;
        col += dcol;
    }
    return spiralMat;
};


let mat1 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
console.log(spiralOrder(mat1));

let mat2 = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]];
console.log(spiralOrder(mat2));

let mat3 = [[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15]];
console.log(spiralOrder(mat3));

File: programming_skills/54_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        row, col, drow, dcol = 0, 0, 0, 1
        spiral_mat = []

        for _ in range(len(matrix) * len(matrix[0])):
            spiral_mat.append(matrix[row][col])
            matrix[row][col] = None
            if (row + drow >= len(matrix) or
                    col + dcol >= len(matrix[0]) or
                    row + drow < 0 or
                    col + dcol < 0 or
                    matrix[row + drow][col + dcol] is None
                ):
                drow, dcol = dcol, -drow
            row += drow
            col += dcol
        return spiral_mat


if __name__ == "__main__":
    matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    solution = Solution()
    print(solution.spiralOrder(matrix))

    matrix = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
    solution = Solution()
    print(solution.spiralOrder(matrix))

    matrix = [[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15]]
    solution = Solution()
    print(solution.spiralOrder(matrix))
